ingestion/src/kldata.rs

```diff
@@ -262,116 +276,131 @@ pub async fn filter_and_label_kldata<'a>(
-                let timeseries_id = transaction
-                    .query_one(
-                        "INSERT INTO public.timeseries (fromtime) VALUES ($1) RETURNING id",
-                        &[&in_datum.timestamp],
-                    )
-                    .await?
-                    .get(0);
+    let mut out_chunks = Vec::with_capacity(chunks.len());
+    for chunk in chunks {
+        let mut data = Vec::with_capacity(chunk.observations.len());
+
+        for in_datum in chunk.observations {
+            // get the conversion first, so we avoid wasting a tsid if it doesn't exist
+            let param = param_conversions
+                .get(&in_datum.id.param_code)
+                .ok_or_else(|| {
+                    Error::Parse(format!(
+                        "unrecognised param_code '{}'",
+                        in_datum.id.param_code
+                    ))
+                })?;
+
+            // TODO: With some changes to this function, we could potentially move its call outside
+            // the loop body. For one thing, the station permit checks, if done in a separate
+            // function, would apply to all observations. Since we know the param-specific permits
+            // are barely used, we could also pre-emptively check all param permits outside the
+            // loop.
+            if !timeseries_is_open(
+                permit_table.clone(),
+                chunk.station_id,
+                chunk.type_id,
+                param.id,
+            )? {
+                // TODO: log that the timeseries is closed? Mostly useful for tests
+                #[cfg(feature = "integration_tests")]
+                eprintln!("station {}: timeseries is closed", chunk.station_id);
+                continue;
+            }
 
-                // create obsinn label
-                transaction
-                    .execute(
-                        "INSERT INTO labels.obsinn \
+            let transaction = conn.transaction().await?;
+
+            let (sensor, lvl) = in_datum
+                .id
+                .sensor_and_level
+                .map(|both| (Some(both.0), Some(both.1)))
+                .unwrap_or((None, None));
+
+            let obsinn_label_result = transaction
+                .query_opt(
+                    &query_get_obsinn,
+                    &[
+                        &chunk.station_id,
+                        &chunk.type_id,
+                        &in_datum.id.param_code,
+                        &lvl,
+                        &sensor,
+                    ],
+                )
+                .await?;
+
+            let timeseries_id: i32 = match obsinn_label_result {
+                Some(row) => row.get(0),
+                None => {
+                    // create new timeseries
+                    // TODO: currently we create a timeseries with null location
+                    // In the future the location column should be moved to the timeseries metadata table
```

Comment:
Will/does lard also support per-obs location? (required for moving sensors)

Reply:
Consensus is that the best way to do this is to have normal timeseries for the latitude and longitude. This is also how the stations treat it, with lat and lon being reported as params. A user would look up a stations location information, see it marked as moving, and so look up its lat and lon params too along with the data they're interested in. APIs like Frost can provide an abstraction over this as appropriate.

Reply:
Does this mean that, for consistency, other obstime-dependent metadata, like quality info are also represented as normal time series? (haven't really checked the db schema, I must admit!) This is fine by me. I guess it does mean that you save space (to the extent that NULLs occupy space) for stationary time series that don't keep changing positions. Maybe some database queries for moving time series would be a little faster if the positions were in the same table as the obs time, but I don't know. Btw, are there any plans of supporting geo search directly in lard (via e.g. PostGIS)?

Reply:
By quality info I mean a one-digit quality code, control info flags (16 digit) from kvalobs, etc. (this is in contrast to performance- and exposure (siting class) info that only rarely change and hence don't need to be kept with the observations). I guess the quality info I'm thinking of is (more or less) what is produced by rove.

If a change of location means a new time series, does that mean that lard makes a distinction between the station location and the obs. location? And that the station location is optional? (think of a sensor mounted in a car - then it doesn't really make sense to keep a station location at all - and as a consequence, there is always only one time series of observations to be generated from that car).

Reply:
> By quality info I mean a one-digit quality code, control info flags (16 digit) from kvalobs,

This is stored alongside the data (`flags.kvdata` table for kvalobs which we will use for the beta, regular data table with the `qc_usable` column for confident)

> If a change of location means a new time series, does that mean that lard makes a distinction between the station location and the obs. location? And that the station location is optional?

We don't currently store any station-level information, we have only the timeseries.

> think of a sensor mounted in a car - then it doesn't really make sense to keep a station location at all

A moving timeseries does not have any location info for the timeseries, location is defined by the lat lon parameters, as discussed above.